### PR TITLE
Add autoconf build system for bundled C sources

### LINKED ISSUE #33: glibc2.36 adds `arc4random_buf`, causing compile error
Per https://sourceware.org/glibc/wiki/Release/2.36, glibc2.36 adds built-in declarations for `arc4random_buf`. When compiling this library as dependency for another project ([ghcup](https://github.com/haskell/ghcup-hs)), the `arc4random_buf` declaration from this project (in [c/archive_random.c](https://github.com/vmchale/libarchive/blob/d81359be55f8fe90fc250bdeef2fc7ab13058db9/c/archive_random.c#L54)) clashes with the built-in declaration, resulting in an error message like so:

```
[15 of 15] Compiling Codec.Archive    ( src/Codec/Archive.hs, dist/build/Codec/Archive.o, dist/build/Codec/Archive.dyn_o )

c/archive_random.c:54:13: error:
     error: static declaration of 'arc4random_buf' follows non-static declaration
       54 | static void arc4random_buf(void *, size_t);
          |             ^~~~~~~~~~~~~~
   |
54 | static void arc4random_buf(void *, size_t);
   |             ^

In file included from c/archive_random.c:30:0: error:

/usr/include/stdlib.h:542:13: error:
     note: previous declaration of 'arc4random_buf' with type 'void(void *, size_t)' {aka 'void(void *, long unsigned int)'}
      542 | extern void arc4random_buf (void *__buf, size_t __size)
          |             ^~~~~~~~~~~~~~
    |
542 | extern void arc4random_buf (void *__buf, size_t __size)
    |             ^
`cc' failed in phase `C Compiler'. (Exit code: 1)
cabal: Failed to build libarchive-3.0.3.2 (which is required by exe:ghcup from
ghcup-0.1.19.0). See the build log above for details.
```

The command being used to build `ghcup` (released version v0.1.19.0) is:
```bash
cabal build --prefix=. --project-file=cabal.project -f+internal-downloader -f+tui -f+system-libarchive
```
(from some testing, this error occurs regardless of whether `-f+system-libarchive` is passed).

I see that there is some way of customizing whether to include this declaration, perhaps by setting `HAVE_ARC4RANDOM_BUF` in [autoconf-linux/config.h](https://github.com/vmchale/libarchive/blob/4e3d6340996a161071dc01c830155f833646a02e/c/autoconf-linux/config.h#L295); however I have no idea how to correctly customize this config when the library is being built automatically by Cabal, as dependency of a different project. Would releasing a new version of this library to be compatible with glibc2.36 help?

### PR DESCRIPTION
**Pull request by hasufell**
Fixes #33

### CODE REVIEW

File: Setup.hs [hunk 1]
```diff
@@ -1,6 +1,41 @@
...
+                                           selectHook >>= \h -> postConf h a c p l
+                         , preBuild    = ((selectHook >>=) .) . flip . flip preBuild
+                         , preCopy     = ((selectHook >>=) .) . flip . flip preCopy
+                         , preClean    = ((selectHook >>=) .) . flip . flip preClean
+                         , preInst     = ((selectHook >>=) .) . flip . flip preInst
+                         , preHscolour = ((selectHook >>=) .) . flip . flip preHscolour
+                         , preHaddock  = ((selectHook >>=) .) . flip . flip preHaddock
+                         , preReg      = ((selectHook >>=) .) . flip . flip preReg
+                         , preUnreg    = ((selectHook >>=) .) . flip . flip preUnreg
+                         }
```
Comment (hasufell) [inline Setup.hs:40]: We detect here whether `system-libarchive` is disabled and in that case use the autoconf hooks and generate the configure script.

File: configure.ac [hunk 2]
```diff
@@ -0,0 +1,383 @@
+m4_define([LIBARCHIVE_VERSION_S],[3.7.2])
```
Comment (hasufell) [inline configure.ac:1]: This file is a partial rip-off of the upstream configure.ac... so it might diverge in the future (just like config.h).

**Comment by hasufell** [comment]
We could also split out into two packages similar to:

* https://hackage.haskell.org/package/lzma
* https://hackage.haskell.org/package/lzma-clib

**Comment by hasufell** [comment]
It seems splitting packages is actually problematic, because `libarchive` itself needs the headers `archive.h` and `archive_entry.h`. I'm not sure cabal has a way of installing headers for a package and having another use them.

**Comment by hasufell** [comment]
> It seems splitting packages is actually problematic, because `libarchive` itself needs the headers `archive.h` and `archive_entry.h`. I'm not sure cabal has a way of installing headers for a package and having another use them.

It actually works with `install-includes`.

**Review by vmchale — APPROVED** [review]
Wonderful, thanks! 

Maybe mention the `autoreconf -fi` step in a `README.md` somewhere? I had to look at the CI script to figure out how to run/test locally.

**Comment by hasufell** [comment]
> Maybe mention the autoreconf -fi step in a README.md somewhere? I had to look at the CI script to figure out how to run/test locally.

Done

**Comment by hasufell** [comment]
Nothing left here I think

**Comment by hasufell** [comment]
@vmchale this is good to go

**Comment by hasufell** [comment]
@vmchale if you make me proper maintainer (including the hackage package, my handle there is @maerwald), then I can finalize this myself.

**Pull request merged**